### ISSUE TITLE
Drones not attacking Frigs

### ISSUE THREAD

**Issue by oldmeat** [ISeeDEDPpl/Questor#371]
Using the most recent build, questor seems to be using drones to attack battleships and crusiers and guns (weapons) on frigs. This behavior is intermittent. Sometimes questor will use both drones and guns on frigs that are scraming/webing but then later in the same pocket drones will work on crusiers/battleships while guns are on a frig. I have also noticed that questor will sometimes continue doing this even when scrammed/webbed.

Since there are gun fits are using heavy long range guns it would be helpful if one could set drones to attack all frigs that have the ship aggo'ed, Then assist primary (gun) target.
It would also be helpful if there was a way to graduate gun targets - say guns kill BS-Frig or Frig-BS.

latest commit Bleeding edge:
2e5de8efc833228eb1a83c360cb0c4c5181e8378

Problem still exist

**Comment by ISeeDEDPpl** [comment]
Did you look at and adjust the ECM priority target settings for primary weapons and drones (separate settings for each flavor of both)
`
   
   
    
  `
 `<addDampenersToPrimaryWeaponsPriorityTargetList>true</addDampenersToPrimaryWeaponsPriorityTargetList>`
`<!-- Use PrimaryWeapons on Targets that use ECMs, if you do not want to make them a primary weapons priority, set this to false. -->`
`<addECMsToPrimaryWeaponsPriorityTargetList>true</addECMsToPrimaryWeaponsPriorityTargetList>`
`<!-- Use PrimaryWeapons on Targets that use Neutralizers, if you do not want to make them a primary weapons priority, set this to false. -->`
`<addNeutralizersToPrimaryWeaponsPriorityTargetList>true</addNeutralizersToPrimaryWeaponsPriorityTargetList>`
`<!-- Use PrimaryWeapons on Targets that use Target Painters, if you do not want to make them a primary weapons priority, set this to false. -->`
  `<addTargetPaintersToPrimaryWeaponsPriorityTargetList>true</addTargetPaintersToPrimaryWeaponsPriorityTargetList`>
 `<!-- Use PrimaryWeapons on Targets that use Tracking Disruptors, if you do not want to make them a primary weapons priority, set this to false. -->`
  `<addTrackingDisruptorsToPrimaryWeaponsPriorityTargetList>true</addTrackingDisruptorsToPrimaryWeaponsPriorityTargetList>`
`<!-- Use PrimaryWeapons on Targets that use Warp Disruptors / Scramblers, if you do not want to make them a primary weapons priority, set this to false. -->`
  `<addWarpScramblersToPrimaryWeaponsPriorityTargetList>true</addWarpScramblersToPrimaryWeaponsPriorityTargetList>`
`<!-- Use PrimaryWeapons on Targets that use Webifiers, if you do not want to make them a primary weapons priority, set this to false. -->`
`<addWebifiersToPrimaryWeaponsPriorityTargetList>true</addWebifiersToPrimaryWeaponsPriorityTargetList>`

`  
   
    
  `
`<addDampenersToDronePriorityTargetList>true</addDampenersToDronePriorityTargetList>`
`<!-- Use PrimaryWeapons on Targets that use ECMs, if you do not want to make them a primary weapons priority, set this to false. -->`
`<addECMsToDroneTargetList>true</addECMsToDroneTargetList>`
`<!-- Use PrimaryWeapons on Targets that use Neutralizers, if you do not want to make them a primary weapons priority, set this to false. -->`
`<addNeutralizersToDronePriorityTargetList>true</addNeutralizersToDronePriorityTargetList>`
`<!-- Use PrimaryWeapons on Targets that use Target Painters, if you do not want to make them a primary weapons priority, set this to false. -->`
`<addTargetPaintersToDronePriorityTargetList>true</addTargetPaintersToDronePriorityTargetList>`
`<!-- Use PrimaryWeapons on Targets that use Tracking Disruptors, if you do not want to make them a primary weapons priority, set this to false. -->`
`<addTrackingDisruptorsToDronePriorityTargetList>true</addTrackingDisruptorsToDronePriorityTargetList>`
`<!-- Use PrimaryWeapons on Targets that use Warp Disruptors / Scramblers, if you do not want to make them a primary weapons priority, set this to false. -->`
`<addWarpScramblersToDronePriorityTargetList>true</addWarpScramblersToDronePriorityTargetList>`
`<!-- Use PrimaryWeapons on Targets that use Webifiers, if you do not want to make them a primary weapons priority, set this to false. -->`
`<addWebifiersToDronePriorityTargetList>true</addWebifiersToDronePriorityTargetList>
`

**Comment by ISeeDEDPpl** [comment]
If you have... and things still seem "off" then you will need to provide a console log (/log/Charactername/Console/DatedLogHere) of an entire pocket that shows the problem. 

I _may_ need a log with debugGetBestTarget==true, I am not sure exactly what you are seeing...

**Comment by oldmeat** [comment]
I have everything set as you wrote and this is log: 
http://pastebin.com/m4d63bnt

I try change all options in   to false, but it does not help

Today i check log with:
debugGetBestTarget==true

**Comment by oldmeat** [comment]
I set debugGetBestTarget==true
and logs look:
[Combat Debug: GetBestTarget:] LowOrHighValueTarget exception [System.NullReferenceException: Object reference not set to an instance of an object.
 in Questor.Modules.Caching.Cache.GetBestTarget(EntityCache currentTarget, Double distance, Boolean lowValueFirst, String callingroutine) w c:\Questor\Questor-BleedingEdge\Questor.Modules\Caching\Cache.cs:line 2956]

**Comment by ISeeDEDPpl** [comment]
OK, I think Ive already corrected this earlier tonight. Can you update to latest bleedingedge and re-test.

OR: go to Questor.Modules\Caching\Cache.cs:line 2956 in your source and paste it in a reply here so I can confirm it is resolved now? (It is HIGHLY likely!)

**Comment by oldmeat** [comment]
I mark line 2956
http://pastebin.com/nWxdAS2H

**Comment by ISeeDEDPpl** [comment]
confirmed: that was fixed in an earlier commit. If you still have issues with latest bleedingedge we will need another log. I assume at this point that it will be working for you.

**Comment by oldmeat** [comment]
Ok, when issue appear again i give fresh logs.

**Comment by oldmeat** [comment]
I send fresh logs to your email

**Comment by ISeeDEDPpl** [comment]
this has been improved quite a bit in bleedingedge. I will need a fresh set of console logs if you have any problems with latest bleedingedge. 

We should no longer shoot frigates with guns (that are too close/ too fast) at all until all other targets have been eliminated. Ive changed both the handling of the priority targets and the choosing of what to actually shoot.

https://github.com/ISeeDEDPpl/Questor/commit/b5ebb4e9cc0fb306021535cfcc3c174fc7bd8afb

**Comment by oldmeat** [comment]
Latest build b5ebb4e:

Still not work properly, primary weapons attack Spider Drone (7,5km) and same time drones attack BS. 
If in pocket are frigs and cruisers, drones always first attack cruisers, not change target.
I send You logs on your email.

**Comment by kn0xv1lle** [comment]
I confirm this behaviour. Watched it a few minutes ago. Still drones attacking bs during primary weapon attack frigates. :/

**Comment by ISeeDEDPpl** [comment]
waiting on logs... the log WilkZly sent did not seem to include the issue. To be safe include the logging for an entire pocket where it occurs please. 

you will need: debugGetBestTarget==true in your characters xml

**Comment by ISeeDEDPpl** [comment]
so much has changed in the last 4 days... this really needs to be re-opened as a new issue with DETAILED logs

i.e.: you will need: debugGetBestTarget==true in your characters xml and you NEED to include the whole pocket (at least). 

I am getting no complaints at the moment so I assume it is working for the majority of cases ok. Please do open a new issue if you can reproduce drones shooting larger targets than they should when smaller targets are avail and/or the same with primary weapons shooting frigates when larger targets are available...

**Comment by ISeeDEDPpl** [comment]
fixed in bleedingedge
